File: sdk/js/src/schemas/enum.ts
import type { ParseContext, SchemaNode } from "../types.js";
import { BaseSchema } from "./base.js";
import { ISSUE_CODES } from "../issue-codes.js";

export class EnumSchema extends BaseSchema<string | number, string | number> {
  private _values: (string | number)[];

  constructor(values: (string | number)[]) {
    super();
    this._values = values;
  }

  _validate(input: unknown, ctx: ParseContext): unknown {
    if (!this._values.includes(input as string | number)) {
      ctx.issues.push({
        code: ISSUE_CODES.INVALID_TYPE,
        message: `Expected one of enum(${this._values.join(",")}), received ${String(input)}`,
        path: [...ctx.path],
        expected: `enum(${this._values.join(",")})`,
        received: String(input),
      });
      return undefined;
    }
    return input;
  }

  _toNode(): SchemaNode {
    const node = {
      kind: "enum" as const,
      values: [...this._values],
    };
    this._addDefault(node as unknown as SchemaNode);
    return node as unknown as SchemaNode;
  }
}

File: sdk/python/src/anyvali/schemas/enum.py
"""Enum schema."""

from __future__ import annotations

from typing import Any

from ..issue_codes import INVALID_TYPE
from .base import BaseSchema, ValidationContext


class EnumSchema(BaseSchema[Any]):
    """Schema that accepts one of a fixed set of values."""

    def __init__(self, values: list[Any]) -> None:
        super().__init__()
        self._values = list(values)

    def _validate(self, input: Any, ctx: ValidationContext) -> Any:
        if input not in self._values:
            values_str = ",".join(str(v) for v in self._values)
            ctx.add_issue(
                INVALID_TYPE,
                f"Expected one of enum({values_str}), received {input!r}",
                expected=f"enum({values_str})",
                received=str(input),
            )
            return None
        # Also check type strictness -- "1" should not match 1
        for v in self._values:
            if v == input and type(v) is type(input):
                return input
        # Type mismatch
        values_str = ",".join(str(v) for v in self._values)
        ctx.add_issue(
            INVALID_TYPE,
            f"Expected one of enum({values_str}), received {input!r}",
            expected=f"enum({values_str})",
            received=str(input),
        )
        return None

    def _to_node(self) -> dict[str, Any]:
        return self._add_common_node_fields({"kind": "enum", "values": list(self._values)})
